gadgets/top_file/test/unit/top_file_test.go

```diff
@@ -0,0 +1,162 @@
+	mnsFilterMap  func(info *utilstest.RunnerInfo) *ebpf.Map
+}
+
+func TestTopFileGadget(t *testing.T) {
+	utilstest.RequireRoot(t)
+	runnerConfig := &utilstest.RunnerConfig{}
+
+	testCases := map[string]testDef{
+		"captures_events_with_filter": {
+			runnerConfig:  runnerConfig,
+			generateEvent: generateEvent,
+			mnsFilterMap: func(info *utilstest.RunnerInfo) *ebpf.Map {
+				return utilstest.CreateMntNsFilterMap(t, info.MountNsID)
+			},
+			validateEvent: func(t *testing.T, info *utilstest.RunnerInfo, filepath string, events []ExpectedTopFileEvent) {
+				utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, pid int) *ExpectedTopFileEvent {
+					return &ExpectedTopFileEvent{
+						Comm: "bash",
+						T:    "R",
+						File: filepath,
+
+						// Only check the existence of pid as bash creates subshell
+						Pid: utils.NormalizedInt,
+						Tid: utils.NormalizedInt,
+
+						// Only check the existence.
+						Writes: utils.NormalizedInt,
+						WBytes: utils.NormalizedInt,
+
+						MntnsID: info.MountNsID,
+						Uid:     0,
+						Gid:     0,
+						Dev:     0,
+
+						// Nothing is being read from the file.
+						Reads:  0,
+						RBytes: 0,
+					}
+				})(t, info, 0, events)
+			},
+		},
+	}
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			var filepath string
+			runner := utilstest.NewRunnerWithTest(t, testCase.runnerConfig)
+			params := map[string]string{
+				"operator.oci.ebpf.map-fetch-interval": "1000ms",
+				"operator.LocalManager.host":           "true",
+			}
+
+			var MnsFilterMap *ebpf.Map
+			if testCase.mnsFilterMap != nil {
+				MnsFilterMap = testCase.mnsFilterMap(runner.Info)
+			}
+			Opts := gadgetrunner.GadgetOpts[ExpectedTopFileEvent]{
+				Image:        "top_file",
+				Timeout:      5 * time.Second,
+				MnsFilterMap: MnsFilterMap,
+				ApiParams:    params,
+			}
+
+			gdgt := gadgetrunner.NewGadget(t, Opts)
+			gdgt.NormalizeEvent = func(event *ExpectedTopFileEvent) {
+				utils.NormalizeInt(&event.Tid)
+				utils.NormalizeInt(&event.Pid)
+				utils.NormalizeInt(&event.WBytes)
+				utils.NormalizeInt(&event.Writes)
+			}
+			gdgt.BeforeGadgetRun = func() error {
+				utilstest.RunWithRunner(t, runner, func() error {
+					filepath = testCase.generateEvent(t)
+					return nil
+				})
+				return nil
+			}
+
+			gdgt.RunGadget()
+
+			testCase.validateEvent(t, runner.Info, filepath, gdgt.CapturedEvents)
+		})
+	}
+}
+func generateEvent(t *testing.T) string {
+	wd, err := os.Getwd()
+	require.NoError(t, err, "getting current working directory")
+
+	filepath := filepath.Join(wd, "bar")
+	file, err := os.Create(filepath)
+	require.NoError(t, err, "creating file")
+
+	defer func() {
+		err := file.Close()
+		require.NoError(t, err, "closing file")
+	}()
+
+	// Set up the command to continuously write to the file with a 5 seconds timeout
+	cmd := exec.Command("bash", "-c", "timeout 5 bash -c 'while true; do echo -n foo > bar; sleep 0.3; done'")
```

Comment:
What about writing the file directly with golang? We could have a very presize control of how many bytes we write and we could check this on the test.

Reply:
I actually tried that out during initially implementing the test, but in my findings the event generated though that approach sometimes gets captured and sometimes act a little flaky, also using it lot of extra events are also generated, bash approach gets captured better, should I shift to using golang?